### PR TITLE
Initial draft for the project design tool

### PR DESCRIPTION
**Pull request by SumayyaZiyad**
$subject

### CODE REVIEW

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# Project Design Diagram
+
+With the increasing size and complexity, software applications have evolved to be a composition of multiple interrelated components. When the number and/or depth of such components grow, it becomes convenient to have a high-level picture of the system components and their interactions with one another. The Ballerina project design view is deployed with the intention of providing a convenient way to generate such a high-level picture of projects, purely based on the source code. In addition to the component interactions, the project design view is also capable of visualizing the Ballerina record types in the project workspace and their associations.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:3]:
```suggestion
With the increasing size and complexity, software applications have evolved to be a composition of multiple interrelated components. When the number and/or depth of such components grow, it becomes convenient to have a high-level picture of the system components and their interactions with one another. The Ballerina Project Design View is deployed with the intention of providing a convenient way to generate such a high-level picture of projects purely based on the source code. In addition to the component interactions, the Project Design View is also capable of visualizing the Ballerina record types in the project workspace and their associations.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+## Set up the prerequisites
+
+Install the latest versions of [Ballerina](https://ballerina.io/downloads/) and [Ballerina Visual Studio Code plugin](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina).
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:7]:
```suggestion
Install the [latest Ballerina version](https://ballerina.io/downloads/) and [Ballerina Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina).
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+The Ballerina Project Design View tool is packed with the latest release of the Ballerina VS Code plugin available in the VS Code marketplace. The diagrams can be generated upon executing the `Ballerina: Project Design` extension command. However, to get the best possible view of the project, it is necessary to setup the Ballerina project with the following concepts in mind.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:9]:
```suggestion
The Ballerina Project Design View tool is shipped with the latest release of the Ballerina VS Code extension available in the VS Code marketplace. The diagrams can be generated upon executing the `Ballerina: Project Design` extension command. However, to get the best possible view of the project, it is necessary to set up the Ballerina project with the following concepts in mind.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+## Related Concepts
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:11]:
```suggestion
## Related concepts
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+### A Ballerina project
+
+In the context of this tool, a Ballerina project is considered to be a collection of one or more Ballerina packages.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:15]:
```suggestion
A Ballerina project is considered to be a collection of one or more Ballerina packages.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+### A project workspace
+
+In the context of this tool, a Ballerina project is contained by a [VS Code workspace](https://code.visualstudio.com/docs/editor/workspaces). Hence, the diagram generator is functional only if the Ballerina packages are added to the IDE as a workspace.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:19]:
```suggestion
A Ballerina project is contained in a [VS Code workspace](https://code.visualstudio.com/docs/editor/workspaces). Hence, the diagram generator is functional only if the Ballerina packages are added to the IDE as a workspace.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+> **Tip:** You can save your VS Code workspace as a file and simply use this file to access the workspace thereafter.
+
+### Service Identifiers
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:22]:
```suggestion
### Service identifiers
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+Since multiple services can have the same path, this tool depends on an additional annotation to uniquely identify the services within the project. The current approach to do so is via the `display` annotation, which is a general-purpose annotation that can be used in Ballerina code.
+
+By providing a unique value to the `id` field of the display annotation, you can assign an identifier to each service. For reference, the `CurrencyService` in the following code snippet has been annotated with an ID of the value `currency`, and this identifier will be used to differentiate the service throughout the project.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:26]:
```suggestion
By providing a unique value to the `id` field of the display annotation, you can assign an identifier to each service. For reference, the `CurrencyService` in the following code snippet has been annotated with an ID of the `currency` value and this identifier will be used to differentiate the service throughout the project.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+   GetSupportedCurrenciesResponse|grpc:Error supportedCurrencies = currencyClient->GetSupportedCurrencies({});
+   if supportedCurrencies is grpc:Error {
+      log:printError("failed to call getSupportedCurrencies from currency service");
+      return supportedCurrencies;
+   }
+   return supportedCurrencies.currency_codes;
+}
+```
+
+### Invoking service resources
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:66]:
```suggestion
### Invoke service resources
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+The resource invocations between services need to be done using [client access actions](https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.2.0#support-for-resource-methods-in-client-objects).
+
+## Explore the `Project Design View` features
+
+For demonstration purposes, this guide will use the [Google Cloud Platform demo](https://github.com/ballerina-guides/gcp-microservices-demo/tree/project-design-demo) to explain the different capabilities of the diagram. This sample has 10 services (9 GRPC and one HTTP) in 10 separate Ballerina packages. The `ui folder` is not a Ballerina package, hence would not make an impact.
+
+As discussed in the [Related Concepts](#a-project-workspace) section, each of these packages need to be loaded in to a VS Code workspace. This could be done using any of the following approaches.
+1. Use the VS Code workspace file that is available in the repository OR
+2. Add each Ballerina package to the workspace using the `Add folder to workspace` feature available VSCode
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:76]:
```suggestion
2. Add each Ballerina package to the workspace using the `Add folder to workspace` feature available in VS Code
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+If the workspace has been set up accurately, you would be able to see the following folder structure in the explorer. However, do note that the workspace name would be different if you set up the workspace from scratch.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:78]:
```suggestion
If the workspace has been set up accurately, you would be able to see the following folder structure in the explorer. 

>**Note** However, the workspace name would be different if you set up the workspace from scratch.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+![Workspace Structure](images/workspace-structure.png)
+
+Once the workspace has been set up, the Ballerina design diagrams generator can be activated via the `Ballerina: Project Design` command. This command can be found in the VS Code command palette as seen below.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:82]:
```suggestion
Once the workspace has been set up, the Ballerina design diagram generator can be activated via the `Ballerina: Project Design` command. This command can be found in the VS Code command palette as shown below.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+![Extension Command](images/extension-command.gif)
+
+The execution of this command will open up a VS Code webview panel that will contain three types of design diagrams. The users can navigate between these diagrams using the side panel available on the diagram webview.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:85]:
```suggestion
The execution of this command will open up a VS Code web-view panel that will contain three types of design diagrams. You can navigate between these diagrams using the side panel available on the diagram web-view.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+### Service Interaction Diagram
+
+The service interaction diagram is the entry point of the project design diagrams tool. This diagram displays all the services available within the workspace with directed links representing any interactions between them. The service interaction diagram generated for the 10 Ballerina packages in the Google Cloud Platform demo project is seen below.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:89]:
```suggestion
The Service Interaction Diagram is the entry point of the Project Design Diagram tool. This diagram displays all the services available within the workspace with directed links representing any interactions between them. The Service Interaction Diagram generated for the 10 Ballerina packages in the Google Cloud Platform demo project is shown below.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+![Service Interaction Diagram](images/service-interaction-diagram.png)
+>**Info:** The directed link between the `Checkout` service and the `Email` service is an indication that a component in the `Checkout` service invokes the `Email` service. In addition to this, the `Email` service also invokes an `external` service via a connector.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:92]:
```suggestion
>**Info:** The directed link between the `Checkout` service and the `Email` service is an indication that a component of the `Checkout` service invokes the `Email` service. In addition to this, the `Email` service also invokes an `external` service via a connector.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+### Detailed Service Interaction Diagram
+The detailed service interaction diagram (Level 2) dives deeper into the compositions of the services and the interactions between them. The following diagram is the detailed representation of the demo project, and it depicts the individual resource/remote functions of the services and their invocations.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:95]:
```suggestion
The detailed Service Interaction Diagram (Level 2) dives deeper into the compositions of the services and the interactions between them. The following diagram is a detailed representation of the demo project, and it depicts the individual resource/remote functions of the services and their invocations.
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+### Types Diagram
+
+The types diagram provides a comprehensive view of all the Ballerina record types defined in the project workspace and their associations.
+
+   ![Types Diagram](images/types-diagram.png)
+
+   >**Info:** The multiplicities of the associations are represented on either side of the connector while the [type inclusions](https://ballerina.io/learn/by-example/type-inclusion-for-records/) are differentiated using a directed link (represents inheritance).
+
+In addition to the above diagrams, the tool also provides the capability to view the **composition** of an individual record type. These composition diagrams can be accessed by:
+   1. Clicking on the record types that are included in the request/response bodies of service invocations shown in the [detailed service diagram](#detailed-service-interaction-diagram)
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:112]:
```suggestion
   1. Clicking on the record types that are included in the request/response bodies of the service invocations shown in the [detailed service diagram](#detailed-service-interaction-diagram)
```

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+   2. Clicking on the head of the records displayed in the types diagram
+
+The following is the composition diagram generated for the `ContextSearchProductsResponse` record type seen in the [types diagram](#types-diagram) above.
+   
+   ![Types Composition Diagram](images/type-composition-diagram.png)
+
+
+### Accessory features
+
+In addition to the above, the Ballerina project design diagram tool also includes the following features.
```
Comment (praneesha) [inline en/docs/architecture-diagram/architecture-diagram.md:122]:
```suggestion
In addition to the above, the Ballerina Project Design Diagram tool also includes the following features.
```

File: en/docs/architecture-diagram/architecture-diagram.md (hunk 13 above)
Comment (SumayyaZiyad) [reply en/docs/architecture-diagram/architecture-diagram.md]: Across VS Code and their documentation, they use `webview` : https://code.visualstudio.com/api/extension-guides/webview

Is it okay if we leave the same?

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+The [Google Cloud Platform demo](https://github.com/ballerina-guides/gcp-microservices-demo/tree/project-design-demo) is used to demonstrate the different capabilities of the diagram. This sample has 10 services (9 GRPC and one HTTP) in 10 separate Ballerina packages. The `ui folder` is not a Ballerina package, hence would not make an impact.
+
+As per the [Related concepts](#a-project-workspace) section, each of these packages needs to be loaded into a VS Code workspace. This could be done using any of the following approaches.
+1. Use the VS Code workspace file that is available in the repository
+2. Add each Ballerina package to the workspace using the `Add folder to workspace` feature available in VS Code
+
+If the workspace has been set up accurately, you would be able to see the following folder structure in the explorer.
+>**Note:** The workspace name would be different if you set up the workspace from scratch.
+
+![Workspace Structure](images/workspace-structure.png)
```
Comment (nilminiwso2) [inline en/docs/architecture-diagram/architecture-diagram.md:81]: Can we move all the images to the docs/img folder [1] and update the links please?

[1] https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/tree/main/en/docs/img

File: en/docs/architecture-diagram/architecture-diagram.md [hunk 20]
```diff
@@ -0,0 +1,126 @@
+# Project Design Diagram
+
+With the increasing size and complexity, software applications have evolved to be a composition of multiple interrelated components. When the number and/or depth of such components grow, it becomes convenient to have a high-level picture of the system components and their interactions with one another. The Ballerina Project Design View is deployed with the intention of providing a convenient way to generate such a high-level picture of projects purely based on the source code. In addition to the component interactions, the project design view is also capable of visualizing the Ballerina record types in the project workspace and their associations.
```
Comment (nilminiwso2) [inline en/docs/architecture-diagram/architecture-diagram.md:3]:
```suggestion
With the increasing size and complexity, software applications have evolved to be a composition of multiple interrelated components. When the number and/or depth of such components grow, it becomes convenient to have a high-level picture of the system components and their interactions with one another. The Ballerina Project Design View is deployed with the intention of providing a convenient way to generate such a high-level picture of projects purely based on the source code. In addition to the component interactions, the Project Design View is also capable of visualizing the Ballerina record types in the project workspace and their associations.
```

**Review by nilminiwso2 — APPROVED** [review]

**Pull request merged**